Add RFC 1003: Mandatory human handle & device label (#5130)
Scille/parsec-cloud: docs/rfcs/1003-mandatory-human-handle.md

## Basic considerations

Human handle field in user certificate is optional.
Device label in device certificate is also optional.

This is due to multiple reasons:

### 1 - Legacy reason

Those fields have been introduced in Parsec v1.13.
Hence all certificates created with an older version lack those fields.

### 2 - Support non-human user reason

The need for human handle appeared from the fact a human wants to be able to
revoke his current account (i.e. his user in Parsec jargon) and re-create one
(typically in case the account is compromised) while retaining his identity
(e.g. user id "Gorge" is already taken by a now revoked user, hence you'll be
known as "Jorge" from now on...).

On the other hand a non-human account are fine with changing there id (typically
revoking `sgx-anti-virus-20230901` and replacing it by `sgx-anti-virus-20230902`).

However in practice we never used non-human accounts.
On top of that the human handle field doesn't prevent from creating non-human account:
the email field can be set to a dummy value (or event to a valid address to contact
the team responsible for this).

### 3 - Hide identity from outsiders

> The Outsider profile is to be use for user that should be allowed to work in

Reviewer: ```suggestion
The Outsider profile is to be used for user that should be allowed to work in
```

an organization while not being trusted enough to know the full identity of

> who is part of said organization.

Reviewer: ```suggestion
who is a part of said organization.
```

Reviewer: Event better <https://www.linguee.fr/francais-anglais/search?source=auto&query=who+is+part+of>

whatsoever).

The email format is divided into `<username>@<domain>`.

For the username part we use the user id field (as email format allow a superset of what
is allowed for `UserID` type).

For the domain part we use the [standard `.invalid` tld](https://en.wikipedia.org/wiki/.invalid).
On top of that we use `redacted.invalid` as domain, considering the only non-legacy reason
for not having a human handle set is because the certificate is a redacted one (in practice
there is very few organization old enough to need legacy support, hence not having a
specific way to handle not-redacted-but-also-no-human-handle-field is not worth it).

### 4 - `redacted.invalid` not allowed for real human handle

Let's consider the following scenario:

1 - User 1 (user id = "alice") is created with an old Parsec version, hence it has no human handle
2 - User 2 (user id = "bob") is created with human handle email = "alice@redacted.invalid"

Now we end up with Parsec displaying the same human handle email for both user 1 and 2 :(
The easy solution for this is simply to consider `redacted.invalid` as a reserved domain that
is not allowed for real human handle.

> Typically `HumanHandle` method would look like:
> ```

Reviewer: ```suggestion
Typically `HumanHandle` method would look like:

```rust
```

impl HumanHandle {
    pub fn new(email: &str, label: &str) -> Result<Self, &'static str> {
      // Create regular human handle from the provided email & label
      // here we ensure email doesn't use `redacted.invalid` domain
    }

    pub fn from_redacted_user(user_id: &UserID) -> Self {
      // Here we compute label and email from the user id (hence email uses
      // `redacted.invalid` domain)
      // This method should only be used in the code deserializing `UserCertificate`
    }
```

### 5 - Don't carry presence/absence of the fields to the GUI

> As we've seen, legacy certificate are handled just like redacted certificate. On top

Reviewer: ```suggestion
As we've seen, legacy certificates are handled just like redacted certificates. On top
```

> of that redacted certificate are only provided to Outsider user.

Reviewer: ```suggestion
of that redacted certificates are only provided to Outsider user.
```

> Given legacy certificate is not that common, there is no need to convey this information

Reviewer: ```suggestion
Given legacy certificates are not that common, there is no need to convey this information
```

to the GUI.

> Regarding redacted certificate, the GUI already has to know the user is an outsider

Reviewer: ```suggestion
Regarding redacted certificates, the GUI already has to know the user is an outsider
```

by the generic error status handler)

Of course, this rule cannot be enforced when loading existing certificates given
they could have been created with Parsec<1.13.
Instead we rely on specific deserialization rules to handle missing human handle /
device label field (see below).
The big pro of dealing at deserialization is we need no specific code to deal with
legacy certificates: they are simply handled as regular redacted certificates.

### 2 - Device label for redacted certificate

Device label is trivial to handle: `DeviceLabel` type is a superset of `DeviceName`,
hence the device's device name (which is always present) can be provided as device label.

In practice what we get:

- Device certificate created with Parsec<1.13: `MyPC` (i.e. device name field)
- Non-redacted certificate: `My PC` (i.e. device label field)
- Redacted certificate: `e5904a3eccd8451b99ba09324104662e` (i.e. device name field, which is always a UUID)

### 3 - Human handle for redacted certificate

Just like for device label, human handle label is a superset of user id so the latter can
be directly used.

For the human handle email we have to be more cunning (especially given the email *is* the

> field used as identifier where label is only displayed to human with no other purpose

Reviewer 2: It is not clear to me why (or where) "the email is the field used as identifier", isn't the user id always use as identifier? 

Which makes me think, do we really need to bother at all with the email's local-part? How about always using something like `redacted@redacted.invalid`?

Reviewer 2: This is indeed interesting for the documentation (not sure if the user documentation though, or maybe dev/admin docs?)

Author: I would say this has it place in the regular: it explain a fundamental concept of Parsec that is useful for advanced end up (the kind of one that read documentation ^^)

typically we want to expose userid/deviceid/realmid in the GUI (in an advanced property section hidden by default) because it helps when working with server administrator

something like `关羽@蜀.汉`).

### 4 - `redacted.invalid` not allowed for real human handle

Let's consider the following scenario:

1 - User 1 (user id = "alice") is created with an old Parsec version, hence it has no human handle
2 - User 2 (user id = "bob") is created with human handle email = "alice@redacted.invalid"

Now we end up with Parsec displaying the same human handle email for both user 1 and 2 :(
The easy solution for this is simply to consider `redacted.invalid` as a reserved domain that
is not allowed for real human handle.

Typically, `HumanHandle` method would look like:

```rust
impl HumanHandle {
    pub fn new(email: &str, label: &str) -> Result<Self, &'static str> {
      // Create regular human handle from the provided email & label
      // here we ensure email doesn't use `redacted.invalid` domain
    }

    pub fn from_redacted_user(user_id: &UserID) -> Self {
      // Here we compute label and email from the user id (hence email uses
      // `redacted.invalid` domain)
      // This method should only be used in the code deserializing `UserCertificate`
    }
```

### 5 - Don't carry presence/absence of the fields to the GUI

> As we've seen, legacy certificates are handled just like redacted certificates. On top
> of that redacted certificates are only provided to Outsider user.
> 
> Given legacy certificates are not that common, there is no need to convey this information
> to the GUI.
> 
> Regarding redacted certificates, the GUI already has to know the user is an outsider
> to adapt what is displayed (e.g. outsider cannot create new workspaces).
> 
> Hence there is no need to provide a specific field to inform where the human handle /
> device label fields come from.

Reviewer 2: It seems to me that "redacted certificates are only provided to Outside user" is no longer true.
I understand what you mean (that this is currently the case), but with this proposed RFC, we will be 
actually switching to "redacted certificates are provided to the Outside users or Standard Legacy users".

Also, some rephrasing to avoid redundancy:
```suggestion
As we've seen, legacy certificates will be handled just like redacted certificates.

However, there is no need to convey this information (redacted or legacy) to the GUI.
Given legacy certificates are not that common, the GUI can adapt what is displayed
only based on user type: 

* Outside user (always redacted certificates): do not show human handle / device information
* Admin/Standard user (regular or redacted certificates): show human handle / device information
  *  If redacted certificate (because legacy user), it is acceptable to show the redacted information in this case

Hence there is no need to provide a specific field to inform where the human handle /
device label fields come from.
```

What this also means is a user can register himself with an email in the `redacted.invalid`

> tld. This is considered ok given the email is validated by the greeter before the creation.

Reviewer 2: Isn't this phrase in contradiction with section 4 (`redacted.invalid not allowed for real human handle`)

<!-- Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS -->

# Mandatory HumanHandle & DeviceLabel fields

## Basic considerations

Human handle field in user certificate is optional.
Device label in device certificate is also optional.

This is due to multiple reasons:

### 1 - Legacy reason

Those fields have been introduced in Parsec v1.13.
Hence all certificates created with an older version lack those fields.

### 2 - Support non-human user reason

The need for human handle appeared from the fact a human wants to be able to
revoke his current account (i.e. his user in Parsec jargon) and re-create one
(typically in case the account is compromised) while retaining his identity
(e.g. user id "Gorge" is already taken by a now revoked user, hence you'll be
known as "Jorge" from now on...).

On the other hand a non-human account are fine with changing there id (typically
revoking `sgx-anti-virus-20230901` and replacing it by `sgx-anti-virus-20230902`).

However in practice we never used non-human accounts.
On top of that the human handle field doesn't prevent from creating non-human account:

> the email field can be set to a dummy value (or event to a valid address to contact

Reviewer: ```suggestion
the email field can be set to a dummy value (or even to a valid address to contact
```

Author: :+1:

the team responsible for this).

### 3 - Hide identity from outsiders

The Outsider profile is to be used for user that should be allowed to work in
an organization while not being trusted enough to know the full identity of
who is part of said organization.

Typical use-case is a contractor working on a specific project (hence only having
access to said project's workspace) while not actually part of the company the
organization belongs to (hence the contractor cannot get the identity of all
person working within the company by dumping the user certificates).

This is achieved by providing the user and device certificates in two flavours:

- regular one

> - redacted: which is the regular one, minus human handle / device label fields
>   which have been set to None

Reviewer: ```suggestion
- redacted: i.e. Regular one without human handle / device label fields
```

It feels simple written this way

## The issue

The human handle / device label fields are almost exclusively used by the GUI (the lower
layers using the user id / device id).

It is annoying for the GUI to have to deal with the fact those fields might be missing,
and in such case to roll a custom logic to still have something to display.

This RFC propose to move the custom logic with the libparsec lower layer, so that
the GUI has the guarantee those fields are always present from it point of view.

## The solution

> tl;dr:

Reviewer: ```suggestion
TL;DR:
```

Author: ![image](https://github.com/Scille/parsec-cloud/assets/3187637/3cf1c3aa-4350-479f-921d-07a2bab70255)

Just like for device label, human handle label is a superset of user id so the latter can
be directly used.

For the human handle email we have to be more cunning (especially given the email *is* the
field used as identifier where label is only displayed to human with no other purpose
whatsoever):

- The user id is case-sensitive while the email is case-insensitive
- The user id accepts unicode characters (encoded as UTF-8) while
  [email address is more messy about it](https://en.wikipedia.org/wiki/International_email)
- The email format is divided into `<username>@<domain>`

For the domain part we use the [standard `.invalid` tld](https://en.wikipedia.org/wiki/.invalid).
On top of that we use `redacted.invalid` as domain, considering the only non-legacy reason
for not having a human handle set is because the certificate is a redacted one (in practice
there is very few organization old enough to need legacy support, hence not having a
specific way to handle not-redacted-but-also-no-human-handle-field is not worth it).

For the username part we cannot use the user id field as-is, but instead apply to it
the `uncaseify` algorithm that encode the case.

`uncaseify` is as follow:

> 1 - replace any `_` by `__`
> 2 - replace any uppercase character by underscore followed by the lowercase version of the character

Reviewer: ```suggestion
1. replace any `_` by `__`
2. replace any uppercase character by underscore followed by the lowercase version of the character
```

https://www.markdownguide.org/basic-syntax/#ordered-lists

Author: :+1:

```python
def uncaseify(txt: str) -> str:
    out = ""
    for c in txt:
        if c.isupper():
            out += "_"
            out += c
        elif c == "_":
            out += "__"
        else:
            out += c
    return out

assert uncaseify("a") == "a"
assert uncaseify("_a_") == "__a__"
assert uncaseify("aAa") == "a_aa"
assert uncaseify("AA") == "_a_a"
assert uncaseify("_A_") == "___a__"
```

`UserID` type has a hard limit of 32 bytes, while the email email address's limit
is 255 bytes (including the domain part and `@`).
The worst case for `uncaseify` is a user id only composed of uppercase (or underscores),
in which case it size is doubled so 64 bytes.
Considering domain is `redacted.invalid` (16 bytes long) and the `@`, there is
238 bytes to accommodate the uncaseified user id (so far more than what is needed).

Finally considering the unicode, the good news is there is nothing to do: our email
format already accept international email address (so a regular user can already use
something like `关羽@蜀.汉`).

### 4 - `redacted.invalid` not allowed for real human handle

Let's consider the following scenario:

> 1 - User 1 (user id = "alice") is created with an old Parsec version, hence it has no human handle
> 2 - User 2 (user id = "bob") is created with human handle email = "alice@redacted.invalid"

Reviewer: ```suggestion
1. User 1 (user id = "alice") is created with an old Parsec version, hence it has no human handle
2. User 2 (user id = "bob") is created with human handle email = "alice@redacted.invalid"
```

Author: :+1:

Now we end up with Parsec displaying the same human handle email for both user 1 and 2 :(
The easy solution for this is simply to consider `redacted.invalid` as a reserved domain that
is not allowed for real human handle.

Typically, `HumanHandle` method would look like:

```rust
impl HumanHandle {
    pub fn new(email: &str, label: &str) -> Result<Self, &'static str> {
      // Create regular human handle from the provided email & label
      // here we ensure email doesn't use `redacted.invalid` domain
    }

>     pub fn from_redacted_user(user_id: &UserID) -> Self {

Reviewer: ```suggestion
    pub fn from_redacted_user(user_id: &UserID) -> Result<Self, &'static str> {
```

Will likely use `HumanHandle::new` since either way we will need to validate the email

Author: We can always create a valid `HumanHandle` from a `UserID`, so no need to return an error

      // Here we compute label and email from the user id (hence email uses
      // `redacted.invalid` domain)

>       // This method should only be used in the code deserializing `UserCertificate`

Reviewer: If this method should only be used by `UserCertificate`, why not making it a method of `UserCertificate` ?

Author: This is also used when deserializing old `LocalDevice` and `LegacyDeviceFilePassword`

on top of that `HumanHandle::new` checks that `redacted.invalid` is not used, hence we need a dedicated method when we need to use it

    }
```

### 5 - Don't carry presence/absence of the fields to the GUI

As we've seen, legacy certificates are handled just like redacted certificates. On top
of that redacted certificates are only provided to Outsider user.

Given legacy certificates are not that common, there is no need to convey this information
to the GUI.

Regarding redacted certificates, the GUI already has to know the user is an outsider
to adapt what is displayed (e.g. outsider cannot create new workspaces).

Hence there is no need to provide a specific field to inform where the human handle /
device label fields come from.

### 6 - Optional device label / human handle turned mandatory in some schemas

Multiple schemas contains an Optional human handle and/or device label.
They can be divided into two categories:

1. the option was present for future-proofing and never used in practice
2. the option was present for backward compatibility

> Items in case 1 are:
> - command `invite_info`

Reviewer: ```suggestion
Items in case 1 are:

- command `invite_info`
```

Author: :+1:

- command `human_find` (removed in APIv4 so can be ignored)
- data `invite_device_confirmation` & `invite_user_confirmation`
- data `invite_device_data` & `invite_user_data`
- data `pki_enrollment_answer_payload`
- `device_file_recovery`
- `device_file_password`
- `device_file_smartcard`

> Items in case 2 are:
> - `legacy_device_file_password`

Reviewer: ```suggestion
Items in case 2 are:

- `legacy_device_file_password`
```

Author: :+1:
